File: ios/Sources/DKMadsSSPSDK/Models/DKMadsNativeAdAssets.swift
import Foundation

/// Parsed native creative fields for custom in-feed layouts (from bid `meta` or root keys).
@objc(DKMadsNativeAdAssets)
public final class DKMadsNativeAdAssets: NSObject {
    @objc public let headline: String?
    @objc public let body: String?
    @objc public let callToAction: String?
    @objc public let advertiser: String?
    @objc public let iconUrl: String?
    @objc public let imageUrl: String?
    @objc public let clickUrl: String?
    /// App/store assets (OpenRTB Native data assets): rating 0–5, price, downloads, likes.
    @objc public let rating: String?
    @objc public let price: String?
    @objc public let downloads: String?
    @objc public let likes: String?

    @objc public init(
        headline: String? = nil,
        body: String? = nil,
        callToAction: String? = nil,
        advertiser: String? = nil,
        iconUrl: String? = nil,
        imageUrl: String? = nil,
        clickUrl: String? = nil,
        rating: String? = nil,
        price: String? = nil,
        downloads: String? = nil,
        likes: String? = nil
    ) {
        self.headline = headline
        self.body = body
        self.callToAction = callToAction
        self.advertiser = advertiser
        self.iconUrl = iconUrl
        self.imageUrl = imageUrl
        self.clickUrl = clickUrl
        self.rating = rating
        self.price = price
        self.downloads = downloads
        self.likes = likes
    }

    @objc public static func from(ad: Ad) -> DKMadsNativeAdAssets {
        ad.nativeAssets
    }

    static func from(dictionary: [String: Any]?) -> DKMadsNativeAdAssets {
        let root = dictionary ?? [:]
        if let assets = root["native_assets"] as? [String: Any] {
            func a(_ keys: [String]) -> String? {
                for key in keys {
                    if let s = assets[key] as? String {
                        let t = s.trimmingCharacters(in: .whitespacesAndNewlines)
                        if !t.isEmpty { return t }
                    } else if let n = assets[key] as? NSNumber {
                        return n.stringValue
                    }
                }
                return nil
            }
            return DKMadsNativeAdAssets(
                headline: a(["headline"]),
                body: a(["body", "description"]),
                callToAction: a(["cta", "cta_label"]),
                advertiser: a(["advertiser"]),
                iconUrl: a(["icon_url"]),
                imageUrl: a(["image_url"]),
                clickUrl: a(["click_url"]),
                rating: a(["rating"]),
                price: a(["price", "saleprice"]),
                downloads: a(["downloads"]),
                likes: a(["likes"])
            )
        }
        let meta = root["meta"] as? [String: Any] ?? [:]
        func str(_ keys: [String]) -> String? {
            for key in keys {
                if let s = (root[key] as? String) ?? (meta[key] as? String) {
                    let t = s.trimmingCharacters(in: .whitespacesAndNewlines)
                    if !t.isEmpty { return t }
                } else if let n = (root[key] as? NSNumber) ?? (meta[key] as? NSNumber) {
                    return n.stringValue
                }
            }
            return nil
        }
        return DKMadsNativeAdAssets(
            headline: str(["headline", "native_title", "title"]),
            body: str(["body", "native_body", "description"]),
            callToAction: str(["cta_label", "call_to_action", "native_cta"]),
            advertiser: str(["advertiser", "sponsored_by", "brand"]),
            iconUrl: str(["icon_url", "native_icon_url"]),
            imageUrl: str(["image_url", "native_image_url", "creativeUrl", "creative_url"]) ?? (root["creativeUrl"] as? String),
            clickUrl: str(["click_url", "clickUrl"]),
            rating: str(["rating"]),
            price: str(["price", "saleprice"]),
            downloads: str(["downloads"]),
            likes: str(["likes"])
        )
    }
}

File: android/DKMadsNativeAdAssets.kt
package com.dkmads.ssp

import org.json.JSONObject

data class DKMadsNativeAdAssets(
    val headline: String? = null,
    val body: String? = null,
    val callToAction: String? = null,
    val advertiser: String? = null,
    val iconUrl: String? = null,
    val imageUrl: String? = null,
    val clickUrl: String? = null,
    /** App/store assets (OpenRTB Native data assets): rating 0–5, price, downloads, likes. */
    val rating: String? = null,
    val price: String? = null,
    val downloads: String? = null,
    val likes: String? = null,
) {
    companion object {
        fun from(ad: Ad): DKMadsNativeAdAssets = ad.nativeAssets ?: DKMadsNativeAdAssets(
            headline = null,
            body = null,
            callToAction = ad.ctaLabel.takeIf { it.isNotBlank() },
            advertiser = null,
            iconUrl = null,
            imageUrl = ad.creativeUrl.takeIf { it.isNotBlank() },
            clickUrl = ad.clickUrl.takeIf { it.isNotBlank() },
        )

        /** Server-provided structured native assets (`winner.native_assets`). */
        fun fromNativeAssets(json: JSONObject): DKMadsNativeAdAssets {
            fun str(key: String): String? = json.optString(key).takeIf { it.isNotBlank() }
            return DKMadsNativeAdAssets(
                headline = str("headline"),
                body = str("body") ?: str("description"),
                callToAction = str("cta") ?: str("cta_label"),
                advertiser = str("advertiser"),
                iconUrl = str("icon_url"),
                imageUrl = str("image_url"),
                clickUrl = str("click_url"),
                rating = str("rating"),
                price = str("price") ?: str("saleprice"),
                downloads = str("downloads"),
                likes = str("likes"),
            )
        }

        fun fromWinner(json: JSONObject): DKMadsNativeAdAssets {
            json.optJSONObject("native_assets")?.let { return fromNativeAssets(it) }
            val meta = json.optJSONObject("meta")
            fun str(vararg keys: String): String? {
                for (key in keys) {
                    val v = json.optString(key).ifBlank { meta?.optString(key).orEmpty().orEmpty() }
                    if (v.isNotBlank()) return v
                }
                return null
            }
            return DKMadsNativeAdAssets(
                headline = str("headline", "native_title", "title"),
                body = str("body", "native_body", "description"),
                callToAction = str("cta_label", "call_to_action", "native_cta"),
                advertiser = str("advertiser", "sponsored_by", "brand"),
                iconUrl = str("icon_url", "native_icon_url"),
                imageUrl = str("image_url", "native_image_url") ?: json.optString("creativeUrl").takeIf { it.isNotBlank() },
                clickUrl = str("click_url", "clickUrl"),
                rating = str("rating"),
                price = str("price", "saleprice"),
                downloads = str("downloads"),
                likes = str("likes"),
            )
        }
    }
}
